Prepare function guidelines (#23291)
prestodb/presto: FUNCTIONS.md

# Presto Function Guidelines

Presto includes a large library of built-in, core SQL functions such as AVG and CONCAT that are available in every deployment.

Presto also includes extension functions that can optionally be enabled in particular deployments.

> Finally, Presto allows users to write user defined functions (UDFs) in either Java or SQL that are not stored in this GitHub repository.

Reviewer: ```suggestion
Finally, Presto allows users to write user defined functions (UDFs) in either Java or SQL that are not stored in this GitHub repository.  [This is done through the Plugin SPI.](https://prestodb.io/docs/current/develop/functions.html)
```

Author: done

This document lays out some guidelines for which category is right for which functions. 

> Functions defined by the ANSI SQL spec should be written as core functions in this repository.

Reviewer 2: Suggest adding a link to the ANSI SQL spec.

Author: Is there such a link? I think you might need to pay for this.

Reviewer 2: If the link costs, then never mind. That said, part of my asking for a link was to ask for a little more detail to help the reader know which ANSI SQL spec (or standard) you are referring them to. Which version is relevant for Presto?  [SQL:2023](https://en.wikipedia.org/wiki/SQL:2023), or one of the other eight versions of the standard?

Author: It's worth specifying that. Do we make any promises about whgich version of ANSI SQL Presto supports?

Reviewer 2: I don't know which version of ANSI SQL Presto supports. If we don't support a specific version, then perhaps "Functions defined by the ANSI SQL spec" does not help the reader know if a function they are working on should be written as a core function and this might need rephrasing or removal. 

@tdcmeehan, is there any guidance you can give as to which if any specific version of ANSI SQL standard Presto supports?

Reviewer: Perhaps we can just generically link to this reference: https://jakewheat.github.io/sql-overview/

Reviewer 2: Okay, that works for me given the context! Added comment with suggested change below.

> Functions defined by the ANSI SQL spec should be written as core functions in this repository.

Reviewer 2: ```suggestion
Functions defined by the [ANSI SQL spec](https://jakewheat.github.io/sql-overview/) should be written as core functions in this repository.
```

Okay, that works for me given the context!

Author: done

Non-standard functions that are commonly available in multiple other SQL dialects such as MySQL and Postgres should also be written as core functions in this library.

Core functions should be broadly applicable to many users in different industries.

> Core functions should be purely computational. That is, they do not make network connections or perform I/O outside the database. (Date/time functions like NOW() are an exception.) For example, a function that talks to a remote server to check the current  temperature or a stock price should not be a core function. This should be a user defined function maintained outside the Presto github repository. 

Reviewer 2: ```suggestion
Core functions should be purely computational. That is, they do not make network connections or perform I/O outside the database. (Date/time functions like NOW() are an exception.) For example, a function that talks to a remote server to check the current temperature or a stock price should not be a core function. This should be a user defined function maintained outside the Presto GitHub repository. 
```

Author: done

Functions that alias or closely resemble existing core functions should not be included in this repository. For instance, because Presto already has an AVG function it should not also have a MEAN or an AVERAGE function, even if the semantics or arguments are slightly different.

> Functions that include trade marked names should be UDFs outside this repository.

Reviewer 2: ```suggestion
Functions that include trademarked names should be UDFs outside this repository.
```

Author: done

Built-in and extension functions should be able to be maintained by Presto committers without support from the original contributors. That is, they should not require specialized or expert knowledge beyond what Presto maintenance already requires. For instance, most cryptographic functions should not be included in this repository because it is not reasonable to expect that Presto committers are also cryptography experts.

Built-in functions that extend Presto into new domains should have a detailed plan and ideally an implementation for a library of functions that cover the domain. For instance, a function that only added US dollars would not be accepted. However, a general library for working with currency data along the lines of [Joda-Money](https://www.joda.org/joda-money/) could be considered. It can be helpful to prototype complex functionality like this in UDFs before writing an RFC and submitting it to the Presto project.

> Built-in and extension functions bundled with Presto must be freely available without fee and compatible with the Presto license. Functions using algorithms that are in any way restricted by copyright, patent, trademark, field of use, or other reasons will not be included in this repository. For instance, a patented video CODEC cannot be used. Functions that include trademarked names should be UDFs outside this repository.

Reviewer: There is no Presto license.

```suggestion
Built-in and extension functions bundled with Presto must be freely available without fee and compatible with an Apache license. Functions using algorithms that are in any way restricted by copyright, patent, trademark, field of use, or other reasons will not be included in this repository. For instance, a patented video CODEC cannot be used. Functions that include trademarked names should be UDFs outside this repository.
```

Author: done

# Presto Function Guidelines

Presto includes a large library of built-in, core SQL functions such as AVG and CONCAT that are available in every deployment.

> Presto also includes extension functions that are not enabled by default but can be optionally enabled in particular deployments.

Author: agreed. Anyone happen to know of any? I'll look in the code, but if anyone can point to one it would save me a bit of time

> Presto also includes extension functions that are not enabled by default but can be optionally enabled in particular deployments.

Reviewer: I believe this is implicit in the fact that functions are a part of the `Plugin` interface.  The way to "enable" functions is to add the plugins.  I don't think there's really any examples of functions that are enabled via a toggle.

> Presto also includes extension functions that are not enabled by default but can be optionally enabled in particular deployments.

Author: rewritten

Finally, Presto allows users to write user defined functions (UDFs) in either Java or SQL that are not stored in this GitHub repository. This is done through the [Plugin SPI](https://prestodb.io/docs/current/develop/functions.html).

This document lays out some guidelines for which category is right for which functions. 

Functions defined in [ANSI SQL](https://jakewheat.github.io/sql-overview/) should be written as core functions in this repository.

Non-standard functions that are commonly available in multiple other SQL dialects such as MySQL and Postgres can also be written as core functions in this library.

> Functions that are provided in order to mirror the syntax and semantics of particular other SQL dialects (MySQL, Postgres, Oracle, etc.) should be extension functions so they can be turned on only by users who need compatibility with that dialect. This also helps avoid conflicts where two existing dialects are incompatible.

Author: yes, that's exactly what made me think of this case. I need to check if they are extension functions.

> Functions that are provided in order to mirror the syntax and semantics of particular other SQL dialects (MySQL, Postgres, Oracle, etc.) should be extension functions so they can be turned on only by users who need compatibility with that dialect. This also helps avoid conflicts where two existing dialects are incompatible.

Author: done
